### PR TITLE
Alpha react-virtualized table

### PR DESCRIPTION
**Pull request by ryantxu**
An alternative to #15811.  This tries: react-virtualized

Styling needs help!  but it works:

image

image

With both libraries, getting reasonable column widths will take work.  `react-table` has more built in table features, while `react-virtualized` is more general.  

Things that need done (i'm not sure if they need done before getting added as an alpha plugin)
1. react based configuration.  Right now you can only test complex thigns by copying over from the angular version :(
1. `isUTC` -- the react plugin needs some way to know if we are in UTC
1. Style cleanup!
1. Test cleanup -- convert the <td> values to <div>
1. Table Transform... I think this should be a general part of #15804

### CODE REVIEW

**Comment by torkelo** [comment]
So having tested both react-table and react-virtualized table which one do you think we should go with? 

Think kibana & chronograf is using react-virtualized. Would need to move the table visualization to @grafana/ui so we can reuse it in explore (and in other panels / plugins). This looks like a great start!

**Comment by ryantxu** [comment]
react-virtualized is a better base library.   react-table has more out of the box features that are handled internally, but in the end we will want direct control anyway so that is actually a disadvantage (ie sorting)

The bulk of this patch is migrating the existing table style configuration blindly.

**Comment by ryantxu** [comment]
@torkelo -- I made some progress moving it to grafana/ui.  That package has much stronger type checking and this is all copied form the untyped existing table code!

What should we name it?  I called it `DataTable`, and it takes an argument `TableData`  😱 

We are using `Table` from react-virtualized, so unless we do some namespace switcheroo that name is taken

**Comment by dprokop** [comment]
@ryantxu does react-table support virtualisation?

**Comment by ryantxu** [comment]
react-table does not support virtualisation -- they promote paging instead :(

**Comment by ryantxu** [comment]
I made a bunch of progress with this, and think it is going in the right direction.  I am using the [MultiGrid](https://bvaughn.github.io/react-virtualized/#/components/MultiGrid) to implement a table.  

The Table component takes grid indexes and then picks a renderer for each cell.  The cell code is super messy port of angular, but I think the interface is reasonable:
```
export interface TableCellBuilderOptions {
  value: any;
  column?: Column;
  row?: any[];
  table?: Table;
  className?: string;
  props: GridCellProps;
}

export type TableCellBuilder = (cell: TableCellBuilderOptions) => ReactElement<'div'>;

/** Simplest cell that just spits out the value */
export const simpleCellBuilder: TableCellBuilder = (cell: TableCellBuilderOptions) => {
  const { props, value, className } = cell;
  const { style } = props;

  return (
    <div style={style} className={'gf-table-cell ' + className}>
      {value}
    </div>
  );
};
```

I have sorting working, and adding fixed columns.  I also added a 'rotate' option that shows swaps row&colum.  This is super useful to show `SELECT LAST(*) FROM X LIMIT 1`

image

The storybook example is fun to play with:
image


The key things that need to happen are:
1. styling cleanup
1. better cell size approach.  I added something that calculates the width/height automatically.  However it logs a message saying it should do *either* width or height, but not both.
1. column style configuration UI
1. The existing ColumStyle > renderer migration.  Lots will need reworked.  In particular I think `filter` options will all need to be rethought
1. fix some odd render issues -- cells sometimes stack?

@torkelo -- how far along does this need to be to get added as alpha?  It would be great to have something that shows table data even if it still needs work before release.  For panels, alpha is clear -- but with storybook I'm not so sure.  I put it in a folder named 'Alpha' :)

File: packages/grafana-ui/src/components/Table/Table.tsx [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+    const rowOffset = showHeader ? -1 : 0;
+
+    if (rotate) {
+      return { column: rowIndex, row: columnIndex + rowOffset };
+    } else {
+      return { column: columnIndex, row: rowIndex + rowOffset };
+    }
+  };
+
+  handleCellClick = (rowIndex: number, columnIndex: number) => {
```
Comment (torkelo) [inline packages/grafana-ui/src/components/Table/Table.tsx]: onCellClick

Comment (ryantxu) [reply packages/grafana-ui/src/components/Table/Table.tsx]: 👍

**Comment by ryantxu** [comment]
@torkelo I think this is ready for alpha.  The base direction/structure is good -- the styles, configuration, and table panel migration still need work, but I think that is appropriate in the alpha.

It would be helpful to have basic table support in react

**Comment by torkelo** [comment]
Yes, will try to find time to review & merge later this week.

**Pull request merged**